Migration from ini config to a Json config

Kantanomo opened the pull request:
Created a class built on top of RapidJson to allow easy access and storage of values inside of a Json File. Class can be used anywhere in the project and is easy to implement anywhere else inside the project.

Main benefits of this change include

- Cleaner settings file with organization of settings.
- Ease of adding new settings to the project
- Code for adding and accessing settings is simplified and has good readability
- Adding support for custom types is very easy

One downside of the update means that we cannot place comments into the config file to give people a description of the settings file.

Kantanomo committed:
initial json config push

Created a class that can be used to access a json file and then get and set values based off their property name and type using template functions.

Kantanomo committed:
remove unessecary template type

Kantanomo committed:
change from template specializations

switched from template specializations to using constexpr statements to detect the types. prevents bulky code while keeping things organized.

Haven't figured out a clean solution to handling char* or std::string. Not sure why they aren't just working as the base behavior of RapidJson is supposed too.

Kantanomo committed:
fixed strings and added realpoint3d

figured out what the issue with strings was, had to use a method StringRef to create an instance of the char* that had a guaranteed lifecycle.

Playing around with adding extra types to the class started with real_point3d. Works as expected.

Kantanomo committed:
Added Nesting and started config upgrading

Added ability to access/create nester objects using the operater[] method inside the class. This allows for organization of the config inside the config file.

Started writing the code to upgrade the old ini to the new one.

Kantanomo committed:
Finished base conversion code

All settings are now transferable to the json format. Next step is auto detection of per-instance configs and flagging the old file as already being upgraded.,

Kantanomo committed:
current pointer cleanup

Removed now uneeded parameter to the function and added some comments to explain the functionality.

Kantanomo committed:
quick github build fix

Issue being caused by dev only settings, full fix will be adding a completely new section just for development settings.

Kantanomo committed:
Finished loading client settings

All client settings now load properly, working on server settings next.

No saving functionality currently exists.

Kantanomo committed:
Server settings loading.

Kantanomo committed:
Code cleanup

Collapsed some duplicate code in the JsonConfig class. Removed old code from the Config namespace functions.

Kantanomo committed:
Config saving done

Kantanomo committed:
Adding missing check for platform type

Kantanomo committed:
Merge branch 'development-new' into Json-Config

Kantanomo committed:
Fixing weird merge error.

Kantanomo committed:
Updating extern name for d3d9ex to config

nukeulater commented on xlive/H2MOD/Modules/Shell/JsonConfig/JsonConfig.cpp:
```diff
@@ -0,0 +1,253 @@
+                return defaultValue;
+        }
+
+    	if constexpr (std::is_same_v<T, bool>) {
```
I think you can remove all of these `constexpr` specializations by simply using `return v.Get<T>();`. And some validation wouldn't hurt because just the type alone doesn't guarantee that the value/object in the json document is of the same type.

Kantanomo replied:
Get<T> had compilation issues, I'll revisit it.

Kantanomo replied:
Merged all of the base RapidJson value types into a single Get<T> call while leaving the expressions for the custom types.

Added Json Value validation that will check if there was a default parameter passed to it or return from the defaultValue store if it doesn't exist. If neither exist it will just return the default value of the specific type.

nukeulater commented on xlive/H2MOD/Modules/Shell/JsonConfig/JsonConfig.cpp:
```diff
@@ -0,0 +1,253 @@
+#include "stdafx.h"
+#include <fstream>
```
move these std libraries to `stdafx.h` if they're not already included there

nukeulater commented on xlive/H2MOD/Modules/Shell/Config.cpp:
```diff
@@ -51,7 +54,7 @@ H2Config_Deadzone_Type H2Config_Controller_Deadzone = H2Config_Deadzone_Type::Ax
 float H2Config_Deadzone_A_Y = 26.0f;
 float H2Config_Deadzone_Radial = 1.0f;
-float H2Config_crosshair_offset = NAN;
+float H2Config_crosshair_offset = 0.138f;
```
default h2v is 0.152 or .156 iirc

Berthalamew replied:
the default is 0.165

Kantanomo replied:
fixed.

Kantanomo committed:
Clean up and changes

Fixed some bad strings.

Fixed an issue with the get function that caused a crash.

Collapsed the base RapidJson types into a single evalutation.

Removed unneeded includes

Kantanomo committed:
Type validation

Added checking to detect if the value inside of the Json file is invalid.

Kantanomo committed:
Fixed weird formatting.

Kantanomo committed:
last commit before making a new branch

making a new branch with the latest commits as merging isn't possible
